src/library/scala/collection/SortedSet.scala

```diff
@@ -29,14 +29,14 @@ trait SortedSet[A] extends Set[A]
 
   override def equals(that: Any): Boolean = that match {
     case _ if this eq that.asInstanceOf[AnyRef] => true
-    case ss: SortedSet[_] if ss.ordering == this.ordering =>
+    case ss: SortedSet[A] if ss.ordering == this.ordering =>
```

Comment:
I'm surprised this doesn't need an `@unchecked` annotation

Reply:
Me too...

```
scala> def f(a: Any) = a match { case s: Set[Int] => 0 }
                                         ^
       warning: non-variable type argument Int in type pattern scala.collection.immutable.Set[Int] (the underlying of Set[Int]) is unchecked since it is eliminated by erasure
def f(a: Any): Int

scala> class C[A] { def f(a: Any) = a match { case s: Set[A] => 0 } }
class C
```

Reply:
is it possible this is a bug/regression? I would swear it used to warn...

Reply:
I'm not sure, but it's worth figuring out. It's used in many places across the collections...

Reply:
This does actually look unsound to me. I may have time to construct a use that leads to a ClassCastException later. I'm not sure what we can do about it.

Reply:
I also have doubts, but couldn't come up with a counterexample...

Reply:
as long as `ss.ordering` has an `equals` method consistent with the contract of equality, I think it should be sound

Reply:
If you allow `equals` to be inconsistent with the contract, it seems you can definitely trigger a ClassCastException. Otherwise, I'm not too sure, my mental code-generation and bytecode checking isn't that reliable.

Reply:
fwiw, when the standard library is compiled with dotty it generates a bunch of unchecked warnings which I don't think anyone has inspected, in this PR this includes the line in question: https://travis-ci.com/github/scala/scala/jobs/497265334

Reply:
I'm hitting "Resolve" because I think this is resolved enough to merge this PR, even if the bigger picture remains unclear.

Reply:
Yeah, definitely the pattern typer letting us down here...

```scala
scala> class C[A](ord: Ordering[A], x: A) { def f(a: Any) = a match { case s: Set[A] => ord.equiv(s.iterator.next(), x) case _ => false } }
class C

scala> new C[String](implicitly[Ordering[String]], "foo").f(Set(1))
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
  at scala.math.Ordering$String$.compare(Ordering.scala:576)
  at scala.math.Ordering.equiv(Ordering.scala:103)
  at scala.math.Ordering.equiv$(Ordering.scala:103)
  at scala.math.Ordering$String$.equiv(Ordering.scala:576)
  at C.f(<console>:1)
  ... 32 elided
```